src/Microsoft.Azure.NotificationHubs/Messaging/AuthorizationRules.cs

```diff
@@ -0,0 +1,301 @@
+        {
+            if (comparand == null)
+            {
+                return false;
+            }
+
+            AuthorizationRuleEqualityComparer equalityComparer = new AuthorizationRuleEqualityComparer();
+            HashSet<AuthorizationRule> thisRules = new HashSet<AuthorizationRule>(this.innerCollection, equalityComparer);
+            HashSet<AuthorizationRule> comparandRules = new HashSet<AuthorizationRule>(comparand.innerCollection, equalityComparer);
+
+            if (thisRules.Count != comparandRules.Count)
+            {
+                return false;
+            }
+
+            foreach (AuthorizationRule rule in thisRules)
+            {
+                if (!comparandRules.Contains(rule))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether [requires encryption].
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if [requires encryption]; otherwise, <c>false</c>.
+        /// </value>
+        public bool RequiresEncryption
+        {
+            get { return this.nameToSharedAccessAuthorizationRuleMap.Any(); }
+        }
+
+        internal void Validate()
+        {
+            foreach (AuthorizationRule rule in this.innerCollection)
+            {
+                rule.Validate();
+            }
+
+            if (this.duplicateAddForSharedAccessAuthorizationRule)
+            {
+                throw new InvalidDataContractException(SRClient.CannotHaveDuplicateSAARule);
+            }
+        }
+
+        internal void UpdateForVersion(ApiVersion version, AuthorizationRules existingAuthorizationRules = null)
+        {
+            if (version < ApiVersion.Three)
+            {
+                // API V2 does not understand SharedAccessAuthorizationRule, so we explicitly
+                // remove it.
+                foreach (var rule in this.nameToSharedAccessAuthorizationRuleMap.Values)
+                {
+                    this.innerCollection.Remove(rule);
+                }
+
+                this.nameToSharedAccessAuthorizationRuleMap.Clear();
+
+                if (existingAuthorizationRules != null)
+                {
+                    // Given an existing rule set (we get existing ruleset in update scenarios), 
+                    // for version 2 and below we need to merge back the SharedAccess rule that 
+                    // is not understood by these version.
+                    foreach (SharedAccessAuthorizationRule rule in existingAuthorizationRules.nameToSharedAccessAuthorizationRuleMap.Values)
+                    {
+                        this.Add(rule);
+                    }
+                }
+            }
+        }
+
+        internal bool IsValidForVersion(ApiVersion version)
+        {
+            if (version < ApiVersion.Three)
+            {
+                if (this.nameToSharedAccessAuthorizationRuleMap.Any())
```

Comment:
Could collapse this to `return !this.nameToSharedAccessAuthorizationRuleMap.Any()`